### PR TITLE
Implement C++ Edge Betweenness Centrality

### PR DESCRIPTION
**Pull request by ChuckHastings**
This PR contains the implementation for edge betweenness centrality.

### CODE REVIEW

File: cpp/src/centrality/betweenness_centrality_impl.cuh [hunk 1]
```diff
@@ -266,6 +288,144 @@ void accumulate_vertex_results(
   }
 }
 
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+void accumulate_edge_results(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+  edge_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, weight_t>& centralities,
```
Comment (seunghwak) [inline cpp/src/centrality/betweenness_centrality_impl.cuh:296]: If you're just changing the array contents, better create a mutable_view and pass the mutable view (https://github.com/rapidsai/cugraph/blob/branch-23.08/cpp/include/cugraph/edge_property.hpp#L134)

File: cpp/src/centrality/betweenness_centrality_impl.cuh [hunk 2]
```diff
@@ -266,6 +288,144 @@ void accumulate_vertex_results(
...
+  rmm::device_uvector<vertex_t>&& distance,
+  rmm::device_uvector<edge_t>&& sigma,
```
Comment (seunghwak) [inline cpp/src/centrality/betweenness_centrality_impl.cuh:298]: `distances` & `sigmas` to better indicate that these are vectors?

File: cpp/src/centrality/betweenness_centrality_impl.cuh [hunk 3]
```diff
@@ -388,10 +549,71 @@ rmm::device_uvector<weight_t> edge_betweenness_centrality(
...
   bool const do_expensive_check)
 {
-  CUGRAPH_FAIL("Not Implemented");
-  // Edge betweenness is computed like vertex betweenness, but you accumulate
-  // centrality on each edge.  We need to adapt this to support edge properties
-  // properly.
+  //
+  // Betweenness Centrality algorithm based on the Brandes Algorithm (2001)
+  //
+  if (do_expensive_check) {}
```
Comment (seunghwak) [inline cpp/src/centrality/betweenness_centrality_impl.cuh:555]: No need to check `vertices_begin, vertices_end` points to valid vertices?

File: cpp/include/cugraph/graph_functions.hpp [hunk 4]
```diff
@@ -350,12 +350,13 @@ void renumber_local_ext_vertices(raft::handle_t const& handle,
...
  * @param graph_view Graph view object of the graph to be decompressed.
+ * @param edge_id_view Optional view object holding edge ids for @p graph_view.
  * @param edge_weight_view Optional view object holding edge weights for @p graph_view.
  * @param renumber_map If valid, return the renumbered edge list based on the provided @p
  * renumber_map
  * @param do_expensive_check A flag to run expensive checks for input arguments (if set to `true`).
- * @return Tuple of edge sources, destinations, and (optional) edge weights (if @p
- * edge_weight_view.has_value() is true).
+ * @return Tuple of edge sources, destinations, (optional) edge ids (if @p edge_id_view.has_value()
+ * is true) and (optional) edge weights (if @p edge_weight_view.has_value() is true).
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/include/cugraph/graph_functions.hpp:359]: Fixed in next push.

File: cpp/src/centrality/betweenness_centrality_impl.cuh [hunk 5]
```diff
@@ -55,6 +59,24 @@ struct brandes_e_op_t {
   }
 };
 
+template <typename vertex_t>
+struct extract_edge_e_op_t {
+  vertex_t d;
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/src/centrality/betweenness_centrality_impl.cuh:64]: Fixed in next push.

File: cpp/src/centrality/betweenness_centrality_impl.cuh (hunk 1 above)
Comment (ChuckHastings) [reply cpp/src/centrality/betweenness_centrality_impl.cuh:296]: Fixed in next push.

File: cpp/src/centrality/betweenness_centrality_impl.cuh (hunk 2 above)
Comment (ChuckHastings) [reply cpp/src/centrality/betweenness_centrality_impl.cuh:298]: Fixed in next push.

File: cpp/src/centrality/betweenness_centrality_impl.cuh (hunk 3 above)
Comment (ChuckHastings) [reply cpp/src/centrality/betweenness_centrality_impl.cuh:555]: Added in next push.

File: cpp/src/centrality/betweenness_centrality_impl.cuh [hunk 6]
```diff
@@ -266,6 +288,144 @@ void accumulate_vertex_results(
...
+
+    update_edge_src_property(
+      handle,
+      graph_view,
+      thrust::make_zip_iterator(distance.begin(), sigma.begin(), delta.begin()),
+      src_properties);
+    update_edge_dst_property(
+      handle,
+      graph_view,
+      thrust::make_zip_iterator(distance.begin(), sigma.begin(), delta.begin()),
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/src/centrality/betweenness_centrality_impl.cuh:424]: We could look at a future optimization that allows the caller to identify a subset of the property tuple to update.  That's a more general purpose issue.

**Review by jnke2016 — APPROVED** [review]
Looks good to me

**Review by seunghwak — APPROVED** [review]
LGTM.

**Pull request merged**